pkg/credentials/filecache.go

```diff
@@ -0,0 +1,230 @@
+	Expiration time.Time
+}
+
+// FileCacheProvider is a file based AWS Credentials Provider implementing expiry and retrieve.
+type FileCacheProvider struct {
+	credentials      *credentials.Credentials // the underlying implementation that has the *real* Provider
+	cachedCredential cachedCredential         // the cached credential, if it exists
+	profile          string
+	clock            Clock
+}
+
+type cacheFile struct {
+	// a map of profiles to cachedCredentials
+	ProfileMap map[string]cachedCredential `yaml:"profiles"`
+}
+
+// Put puts the given cachedCredential with a given key into the map. It will overwrite
+// if the key already exists.
+func (c *cacheFile) Put(key string, credential cachedCredential) {
+	c.ProfileMap[key] = credential
+}
+
+// Get returns cachedCredential if it exists in the cred store.
+func (c *cacheFile) Get(key string) cachedCredential {
+	var credential cachedCredential
+	if _, ok := c.ProfileMap[key]; ok {
+		credential = c.ProfileMap[key]
+	}
+	return credential
+}
+
+// NewFileCacheProvider creates a new filesystem based AWS credential cache. The cache uses Expiry provided by the
+// AWS Go SDK for providers. It wraps the configured credential provider into a file based cache provider. If the provider
+// does not support caching ( I.e.: it doesn't implement IsExpired ) then this file based caching system is ignored
+// and the default credential provider is used. Caches are per profile.
+func NewFileCacheProvider(profile string, creds *credentials.Credentials, clock Clock) (FileCacheProvider, error) {
+	if creds == nil {
+		return FileCacheProvider{}, errors.New("no underlying Credentials object provided")
+	}
+	filename, err := cacheFilename()
+	if err != nil {
+		return FileCacheProvider{}, fmt.Errorf("failed to get cache file: %w", err)
+	}
+	if err := os.MkdirAll(filepath.Dir(filename), 0700); err != nil {
+		return FileCacheProvider{}, fmt.Errorf("failed to create folder: %w", err)
+	}
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		logger.Warning("Cache file %s does not exist.\n", filename)
+		return FileCacheProvider{
+			profile:          profile,
+			credentials:      creds,
+			cachedCredential: cachedCredential{},
+			clock:            clock,
+		}, nil
+	}
+
+	if info.Mode()&0077 != 0 {
+		// cache file has secret credentials and should only be accessible to the user, refuse to use it.
+		return FileCacheProvider{}, fmt.Errorf("cache file %s is not private", filename)
+	}
+
+	cache, err := readCacheFile(filename)
+	if err != nil {
+		return FileCacheProvider{}, err
+	}
+
+	return FileCacheProvider{
+		credentials:      creds,
+		cachedCredential: cache.Get(profile),
+		profile:          profile,
+		clock:            clock,
+	}, nil
+}
+
+// readCacheFile reads the contents of the credential cache and returns the
+// parsed yaml as a cachedCredential object.
+func readCacheFile(filename string) (cacheFile, error) {
+	lock := flock.New(filename)
+	defer func() {
+		_ = lock.Unlock()
```

Comment:
do we want to log or smth if unlocking fails? 🤔 i know we will find out that the lock is still claimed the next time it is called, but could be helpful 🤷‍♀️

Reply:
Sure, I can add some log lines

Reply:
Added :)

Reply:
is that message correct?

Reply:
Now it is. :D :D

Reply:
You know, reading really helped further your carrier.